### PR TITLE
[API] Add LDAP auth method

### PR DESCRIPTION
**Pull request by RickyGrassmuck**
This commit adds LDAP to the available Vault API auth methods.

Recreated to replace #13465

### CODE REVIEW

File: api/auth/ldap/ldap.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+var _ api.AuthMethod = (*LDAPAuth)(nil)
+
+type LoginOption func(a *LDAPAuth) error
+
+const (
+	defaultMountPath = "ldap"
+)
+
+// NewLDAPAuth initializes a new Userpass auth method interface to be
```
Comment (digivava) [inline api/auth/ldap/ldap.go:29]: Copy-paste typo here, I think you probably meant "a new LDAP auth method".

File: api/auth/ldap/ldap.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+type LDAPAuth struct {
+	mountPath string
+	username  string
+	password  string
+}
+
+type Password struct {
+	// The password as a plaintext string value.
+	FromString string
+}
```
Comment (digivava) [inline api/auth/ldap/ldap.go:19]: With the Userpass and Approle auth modules, we provide the ability to pass in the password FromString, FromEnv, or FromFile, just to allow for different use cases (and to keep consistent). 

Would you be able to provide the same thing here, or is there a reason you only chose to use FromString?

**Review by digivava — CHANGES_REQUESTED** [review]
Thank you for the contribution to this! I've just mentioned one thing you might want to add and then otherwise it seems good to go.

File: api/auth/ldap/ldap.go (hunk 2 above)
Comment (RickyGrassmuck) [reply api/auth/ldap/ldap.go:27]: I think, at the time, my reasoning was likely centered around this being used in more interactive scenarios given the nature of LDAP authentication.

After reading this, It reminded me that I actually have some projects I plan to use this in where having `FromFile` and `FromEnv` would be necessary so I'm positive others out there would benefit from it. 

Got those two methods added in as well as their test coverage, let me know if there's anything else!

File: api/auth/ldap/ldap.go [hunk 3]
```diff
@@ -81,3 +116,46 @@ func WithMountPath(mountPath string) LoginOption {
...
+	}
+
+	passwordValue := strings.TrimSuffix(string(passwordBytes), "\n")
+
+	return passwordValue, nil
+}
+
+func (password *Password) validate() error {
+	if password.FromFile == "" && password.FromEnv == "" && password.FromString == "" {
+		return fmt.Errorf("password for Userpass auth must be provided with a source file, environment variable, or plaintext string")
```
Comment (digivava) [inline api/auth/ldap/ldap.go:140]: Appreciate it! Just one more Userpass -> LDAP change that needs to happen here and then we’re good.

Comment (RickyGrassmuck) [reply api/auth/ldap/ldap.go]: All set!

**Review by digivava — APPROVED** [review]
Great! Thanks again!

**Pull request merged**